src/pages/api/reviews/[movieId]/sendReview/index.ts

```diff
@@ -0,0 +1,41 @@
+import connectMongo from "../../../../../util/connectMongo";
+import Movie from "../../../../../../models/movie";
+
+/**
+ * @param {import('next').NextApiRequest} req
+ * @param {import('next').NextApiResponse} res
+ */
+
+export default async function createReview(req: any, res: any) {
+  try {
+    const { movieId } = req.query;
+
+    await connectMongo();
+
+    // Finds movie by title
+    const movie = await Movie.findOne({ title: movieId });
+
+    // if movieId dont match
+    if (!movie) {
+      return res.status(404).json({ error: "Movie not found" });
+    }
+
+    // New review object
+    const newReview = {
+      reviewerName: req.body.reviewerName,
+      reviewerText: req.body.reviewerText,
+      postDate: new Date(),
+      rating: req.body.rating,
+    };
+
+    // Add the new review to reviews array
+    movie.reviews.push(newReview);
+
+    // Saves
+    await movie.save();
```

Comment:
You have a situation here that can cause a _race condition_. A race condition is when two users of some code executes the code simultaneously, and only one of them will have effect.

In this case, if two users A and B both submit reviews simultaneously to a movie that have no review previously, the way this code will execute looks like this.

1. Code being executed for user A loads the movie (`await Movei.findOne()` on line 16) and while it loads, surrenders control to the server (through async programming, `await`). This returns the movie with no reviews.
2. Same thing happens for user B. This also returns the movie with no reviews.
3. Later in the execution for A, the new review from A is added to the empty list of reviews (line 32) and the movie is then saved to the database (link 35)
4. Finally, in the execution for B, the new review from B is added to the empty list, because the list is still empty (nothing has happened to it in this function since it was loaded on line 16)

The end result is that only the review from user B will be stored, overwriting the one from user A.

This is related to the ACID principles of database programming. The way mongodb solves this is by providing specialized functions for updating objects (e.g. the movie) like this. 

Here's a StackOverflow thread about this topic: https://stackoverflow.com/questions/33049707/push-items-into-mongo-array-via-mongoose